### ISSUE TITLE
gcsfuse help is pretty ugly

### ISSUE THREAD

**Issue by jgeewax** [GoogleCloudPlatform/gcsfuse#89]
### How gcsfuse shows help

```
jjg@jjg1:~/projects$ gcsfuse
Usage: gcsfuse [flags] bucket_name mount_point

Flags:
  -dir-mode=493: Permissions bits for directories. Default is 0755.
  -file-mode=420: Permissions bits for files. Default is 0644.
  -fuse.debug=false: Write FUSE debugging messages to stderr.
  -fuse.trace_by_pid=false: Enable a hacky mode that uses reqtrace to group all ops from each individual PID. Not a good idea to use in production; races, bugs, and resource leaks likely lurk.
  -fuseutil.random_delays=false: If set, randomly delay each op received, to help expose concurrency issues.
  -gcs-chunk-size=16777216: If set to a non-zero value N, split up GCS objects into multiple chunks of size at most N when reading, and do not read or cache unnecessary chunks.
  -gcs.debug=false: Write FUSE debugging messages to stderr.
  -gid=-1: If non-negative, the GID that owns all inodes. The default is the GID of the gcsfuse process.
  -httputil.debug=false: Dump information about HTTP requests.
  -implicit-dirs=false: Implicitly define directories based on their content. See docs/semantics.md.
  -limit-bytes-per-sec=-1: If positive, a limit on the GCS -> gcsfuse bandwidth for reading objects, measured over a 30-second window.
  -limit-ops-per-sec=5: If positive, a limit on the rate at which we send requests to GCS, measured over a 30-second window.
  -o=map[]: Additional system-specific mount options. Be careful!
  -reqtrace.enable=false: Collect and print traces.
  -stat-cache-ttl=1m0s: How long to cache StatObject results from GCS.
  -syncutil.check_invariants=false: Crash when registered invariants are violated.
  -temp-dir="": The temporary directory in which to store local copies of GCS objects. If empty, the system default (probably /tmp) will be used.
  -temp-dir-bytes=2147483648: A desired limit on the number of bytes used in --temp-dir. May be exceeded for dirty files that have not been flushed or closed.
  -type-cache-ttl=1m0s: How long to cache name -> file/dir type mappings in directory inodes.
  -uid=-1: If non-negative, the UID that owns all inodes. The default is the UID of the gcsfuse process.
```

### How a typical binary shows help

```
jjg@jjg1:~/projects$ cat --help
Usage: cat [OPTION]... [FILE]...
Concatenate FILE(s), or standard input, to standard output.

  -A, --show-all           equivalent to -vET
  -b, --number-nonblank    number nonempty output lines, overrides -n
  -e                       equivalent to -vE
  -E, --show-ends          display $ at end of each line
  -n, --number             number all output lines
  -s, --squeeze-blank      suppress repeated empty output lines
  -t                       equivalent to -vT
  -T, --show-tabs          display TAB characters as ^I
  -u                       (ignored)
  -v, --show-nonprinting   use ^ and M- notation, except for LFD and TAB
      --help     display this help and exit
      --version  output version information and exit

With no FILE, or when FILE is -, read standard input.

Examples:
  cat f - g  Output f's contents, then standard input, then g's contents.
  cat        Copy standard input to standard output.

Report cat bugs to bug-coreutils@gnu.org
GNU coreutils home page: <http://www.gnu.org/software/coreutils/>
General help using GNU software: <http://www.gnu.org/gethelp/>
For complete documentation, run: info coreutils 'cat invocation'
```

### Key issues to fix

- [ ] No short summary of what the binary does (`cat` has "Concatenate FILE(s), or standard input, to standard output.")
- [x] No proper spacing
- [x] No short and long names (`cat` has `-A, --show-all`)
- [x] Long names use underscores instead of hyphens (`bucket_name` instead of `bucket-name` or just `bucket`)
- [ ] Dots in parameters (`fuse.debug`) ...?
- [x] Defaults list as `=` instead of `Default: ...`
- [ ] Lower-case placeholders (`[flags]` instead of `[FLAGS]`)
- [ ] No examples (ie, `cat f - g  Output f's contents, then standard input, then g's contents.`)
- [ ] No end summary (where to report bugs, links to pages, etc)

**Comment by jgeewax** [comment]
For further comparison, here's how `sshfs` shows help (note the different sections for the various options):

```
jjg@jjg1:~/projects$ sshfs --help
usage: sshfs [user@]host:[dir] mountpoint [options]

general options:
    -o opt,[opt...]        mount options
    -h   --help            print help
    -V   --version         print version

SSHFS options:
    -p PORT                equivalent to '-o port=PORT'
    -C                     equivalent to '-o compression=yes'
    -F ssh_configfile      specifies alternative ssh configuration file
    -1                     equivalent to '-o ssh_protocol=1'
    -o reconnect           reconnect to server
    -o delay_connect       delay connection to server
    -o sshfs_sync          synchronous writes
    -o no_readahead        synchronous reads (no speculative readahead)
    -o sync_readdir        synchronous readdir
    -o sshfs_debug         print some debugging information
    -o cache=BOOL          enable caching {yes,no} (default: yes)
    -o cache_timeout=N     sets timeout for caches in seconds (default: 20)
    -o cache_X_timeout=N   sets timeout for {stat,dir,link} cache
    -o workaround=LIST     colon separated list of workarounds
             none             no workarounds enabled
             all              all workarounds enabled
             [no]rename       fix renaming to existing file (default: off)
             [no]nodelaysrv   set nodelay tcp flag in sshd (default: off)
             [no]truncate     fix truncate for old servers (default: off)
             [no]buflimit     fix buffer fillup bug in server (default: on)
    -o idmap=TYPE          user/group ID mapping, possible types are:
             none             no translation of the ID space (default)
             user             only translate UID of connecting user
             file             translate UIDs/GIDs contained in uidfile/gidfile
    -o uidfile=FILE        file containing username:remote_uid mappings
    -o gidfile=FILE        file containing groupname:remote_gid mappings
    -o nomap=TYPE          with idmap=file, how to handle missing mappings
             ignore           don't do any re-mapping
             error            return an error (default)
    -o ssh_command=CMD     execute CMD instead of 'ssh'
    -o ssh_protocol=N      ssh protocol to use (default: 2)
    -o sftp_server=SERV    path to sftp server or subsystem (default: sftp)
    -o directport=PORT     directly connect to PORT bypassing ssh
    -o slave               communicate over stdin and stdout bypassing network
    -o disable_hardlink    link(2) will return with errno set to ENOSYS
    -o transform_symlinks  transform absolute symlinks to relative
    -o follow_symlinks     follow symlinks on the server
    -o no_check_root       don't check for existence of 'dir' on server
    -o password_stdin      read password from stdin (only for pam_mount!)
    -o SSHOPT=VAL          ssh options (see man ssh_config)

FUSE options:
    -d   -o debug          enable debug output (implies -f)
    -f                     foreground operation
    -s                     disable multi-threaded operation

    -o allow_other         allow access to other users
    -o allow_root          allow access to root
    -o auto_unmount        auto unmount on process termination
    -o nonempty            allow mounts over non-empty file/dir
    -o default_permissions enable permission checking by kernel
    -o fsname=NAME         set filesystem name
    -o subtype=NAME        set filesystem type
    -o large_read          issue large read requests (2.4 only)
    -o max_read=N          set maximum size of read requests

    -o hard_remove         immediate removal (don't hide files)
    -o use_ino             let filesystem set inode numbers
    -o readdir_ino         try to fill in d_ino in readdir
    -o direct_io           use direct I/O
    -o kernel_cache        cache files in kernel
    -o [no]auto_cache      enable caching based on modification times (off)
    -o umask=M             set file permissions (octal)
    -o uid=N               set file owner
    -o gid=N               set file group
    -o entry_timeout=T     cache timeout for names (1.0s)
    -o negative_timeout=T  cache timeout for deleted names (0.0s)
    -o attr_timeout=T      cache timeout for attributes (1.0s)
    -o ac_attr_timeout=T   auto cache timeout for attributes (attr_timeout)
    -o noforget            never forget cached inodes
    -o remember=T          remember cached inodes for T seconds (0s)
    -o intr                allow requests to be interrupted
    -o intr_signal=NUM     signal to send on interrupt (10)
    -o modules=M1[:M2...]  names of modules to push onto filesystem stack

    -o max_write=N         set maximum size of write requests
    -o max_readahead=N     set maximum readahead
    -o max_background=N    set number of maximum background requests
    -o congestion_threshold=N  set kernel's congestion threshold
    -o async_read          perform reads asynchronously (default)
    -o sync_read           perform reads synchronously
    -o atomic_o_trunc      enable atomic open+truncate support
    -o big_writes          enable larger than 4kB writes
    -o no_remote_lock      disable remote file locking
    -o no_remote_flock     disable remote file locking (BSD)
    -o no_remote_posix_lock disable remove file locking (POSIX)
    -o [no_]splice_write   use splice to write to the fuse device
    -o [no_]splice_move    move data while splicing to the fuse device
    -o [no_]splice_read    use splice to read from the fuse device

Module options:

[iconv]
    -o from_code=CHARSET   original encoding of file names (default: UTF-8)
    -o to_code=CHARSET	    new encoding of the file names (default: UTF-8)

[subdir]
    -o subdir=DIR	    prepend this directory to all paths (mandatory)
    -o [no]rellinks	    transform absolute symlinks to relative
```

I'd expect us to have:

1. general options
1. auth options
1. GCS options
1. FUSE options

**Comment by jgeewax** [comment]
Some notes

- http://gerg.ca/blog/post/2012/go-command-line/ has a bunch of options
- https://github.com/codegangsta/cli is also not on that list, and works pretty well

**Comment by jgeewax** [comment]
Some others that might be worth looking at:

- https://github.com/alecthomas/kingpin
- https://github.com/jessevdk/go-flags

Kingping might be a good fit...

**Comment by jacobsa** [comment]
Thanks for investigating. It looks like you're hacking on this; should I assign it to you?

**Comment by jgeewax** [comment]
Sure. There's an open pull request for you to review.

**Comment by jacobsa** [comment]
Oops, out of order email. Sorry.

**Comment by jacobsa** [comment]
As discussed in #91, these flags are not controlled by gcsfuse but rather its dependencies:

* [`fuse.debug`](https://github.com/jacobsa/fuse/blob/069fff34a5ce9adba098672b60e84ffdb3056522/debug.go#L26-L29)
* [`fuse.trace_by_pid`](https://github.com/jacobsa/fuse/blob/069fff34a5ce9adba098672b60e84ffdb3056522/connection.go#L34-L39)
* [`fuseutil.random_delays`](https://github.com/jacobsa/fuse/blob/069fff34a5ce9adba098672b60e84ffdb3056522/fuseutil/file_system.go#L28-L30)
* [`gcs.debug`](https://github.com/jacobsa/gcloud/blob/8289aa9f1290a6dead04fa77e9c29b7f0fe74cb7/gcs/debug_bucket.go#L30-L33)
* [`httputil.debug`](https://github.com/jacobsa/gcloud/blob/8289aa9f1290a6dead04fa77e9c29b7f0fe74cb7/httputil/debugging_round_tripper.go#L26-L29)
* [`reqtrace.enable`](https://github.com/jacobsa/reqtrace/blob/e4daabf6acca9c74b7efc8d786251336313c829e/reqtrace.go#L26)
* [`syncutil.check_invariants`](https://github.com/jacobsa/syncutil/blob/47ae5379a1cd4265867a8748ed01355df33e91a7/invariant_mutex.go#L22)

Will need to clean up each one, then adopt @jgeewax's change from #91, fix it up so it can be merged cleanly, then merge it.

**Comment by jacobsa** [comment]
Better order:

1. Pull in #91.
2. Merge in `master`.
3. Fix outstanding issues mentioned in #91 aside from these flags.
4. One by one, define new flags for these and inject their values into the dependencies.
